tests/benchmark/crud_test.go

```diff
@@ -0,0 +1,920 @@
+// Copyright 2022 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package benchmark
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/apigee/registry/connection"
+	"github.com/apigee/registry/rpc"
+	"google.golang.org/genproto/protobuf/field_mask"
+)
+
+func getRootResource(b *testing.B) string {
+	b.Helper()
+
+	p, ok := os.LookupEnv("REGISTRY_PROJECT_IDENTIFIER")
+	if !ok || p == "" {
+		p = "bench"
+	}
+	return fmt.Sprintf("projects/%s/locations/global", p)
+}
+
+func loadVersionCount(b *testing.B) int {
+	b.Helper()
+
+	s, ok := os.LookupEnv("REGISTRY_VERSION_COUNT")
+	if !ok {
+		return 1
+	}
+
+	c, err := strconv.Atoi(s)
+	if err != nil {
+		b.Fatalf("Setup: Invalid REGISTRY_VERSION_COUNT %q: count must be an integer", s)
+	}
+
+	return c
+}
+
+func getApiName(i int, b *testing.B) string {
+	b.Helper()
+
+	prefix, ok := os.LookupEnv("API_NAME_PREFIX")
+	if !ok {
+		prefix = "benchtest"
+	}
+	s, ok := os.LookupEnv("API_NAME_START_OFFSET")
+	if ok {
+		offset, err := strconv.Atoi(s)
+		if err != nil {
+			b.Fatalf("Setup: Invalid API_NAME_START_OFFSET %q: offset must be an integer", s)
+		}
+		i += offset
+	}
+
+	return fmt.Sprintf("%s-%d", prefix, i)
+}
+
+func readAndGZipFile(filename string) (*bytes.Buffer, error) {
+	fileBytes, _ := ioutil.ReadFile(filename)
+	var buf bytes.Buffer
+	zw, _ := gzip.NewWriterLevel(&buf, gzip.BestCompression)
+	_, err := zw.Write(fileBytes)
+	if err != nil {
+		return nil, err
+	}
+	if err := zw.Close(); err != nil {
+		return nil, err
+	}
+	return &buf, nil
+}
+func pauseAfter1000Iterations(i int) {
+	if i%1000 == 0 {
+		/*
+			Pause 5 seconds every 1000 iterations to reduce chances of failure on
+			CREATE, UPDATE, DELETE
+		*/
+		time.Sleep(5 * time.Second)
+	}
+}
```

Comment:
Can we replace this with a simpler throttle? Making a file-level variable with a fixed duration to sleep between _every_ iteration seems much easier to understand.

Reply:
I made it a parameter that can be passed.